### PR TITLE
Fixes #23027: Pin config context cache invalidation to the saving database connection

### LINKED ISSUE #23027: Config context cache invalidation ignores the saving database connection
### NetBox Edition

NetBox Community

### NetBox Version

v4.7.0-beta1

### Python Version

3.14

### Steps to Reproduce

The config-context cache invalidation added in #21025 is new in 4.7; this does not affect 4.6.x.

1. Add a temporary test module at `netbox/extras/tests/test_repro.py`:

```python
from django.test import TestCase, override_settings

from dcim.models import Device, DeviceRole, DeviceType, Location, Manufacturer, Site
from utilities.testing import PinnedConnectionRouter
from virtualization.models import Cluster, ClusterType, VirtualMachine


class TmpRepro(TestCase):

    def test_location_site_change_routes_device_query(self):
        site_a = Site.objects.create(name='Site A', slug='site-a')
        site_b = Site.objects.create(name='Site B', slug='site-b')
        location = Location.objects.create(name='Loc', slug='loc', site=site_a)
        manufacturer = Manufacturer.objects.create(name='M', slug='m')
        Device.objects.create(
            name='Device',
            site=site_a,
            location=location,
            device_type=DeviceType.objects.create(manufacturer=manufacturer, model='DT'),
            role=DeviceRole.objects.create(name='R', slug='r'),
        )

        location = Location.objects.get(pk=location.pk)
        location.site_id = site_b.pk
        with override_settings(DATABASE_ROUTERS=[PinnedConnectionRouter(Device)]):
            location.save()

    def test_cluster_scope_change_routes_vm_query(self):
        site_a = Site.objects.create(name='Site A', slug='site-a')
        site_b = Site.objects.create(name='Site B', slug='site-b')
        cluster_type = ClusterType.objects.create(name='CT', slug='ct')
        cluster = Cluster.objects.create(name='C', type=cluster_type, scope=site_a)
        VirtualMachine.objects.create(name='VM', cluster=cluster)

        cluster = Cluster.objects.get(pk=cluster.pk)
        cluster.scope_id = site_b.pk
        with override_settings(DATABASE_ROUTERS=[PinnedConnectionRouter(VirtualMachine)]):
            cluster.save()
```

2. Run it:

```
export NETBOX_CONFIGURATION=netbox.configuration_testing
python manage.py test extras.tests.test_repro
```

`PinnedConnectionRouter` (`utilities/testing/utils.py`) raises on any read or write of the listed models that is not pinned to an explicit database alias. Django consults `DATABASE_ROUTERS` only for queries that name no connection, so a handler that threads through the alias supplied by the `post_save` signal never reaches the router.

### Expected Behavior

Both tests pass. Every query the config-context invalidation makes should be issued against the connection the triggering object was saved on — the alias Django passes to the `post_save`/`m2m_changed`/`pre_delete` receiver as `using` — as the scope propagation handlers in `dcim.signals` and `virtualization.signals` do since #22922 and #22963.

### Observed Behavior

Both tests error:

```
ERROR: test_location_site_change_routes_device_query
  File "netbox/extras/signals.py", line 261, in _handler
  File "netbox/extras/cache.py", line 29, in invalidate_config_context_for_objects
utilities.testing.utils.UnpinnedQuery: Device query was routed rather than pinned to a connection

ERROR: test_cluster_scope_change_routes_vm_query
  File "netbox/extras/signals.py", line 266, in _handler
  File "netbox/extras/cache.py", line 29, in invalidate_config_context_for_objects
utilities.testing.utils.UnpinnedQuery: VirtualMachine query was routed rather than pinned to a connection
```

On a deployment with `DATABASE_ROUTERS` configured (e.g. `netbox_branching`), a router that resolves these to a different alias would read the affected PKs from — and NULL `_config_context_data` in — a database other than the one holding the change that triggered the invalidation. The intended objects keep a stale pre-rendered config context, and unrelated objects in the other database are needlessly invalidated. The write also falls outside the transaction the save runs in, so it is not rolled back with a failed save.

### Suspected Cause

Neither `extras/cache.py` nor the receivers in `extras/signals.py` accept or forward the signal's `using` alias, so every query in the invalidation path falls through to the router.

`extras/cache.py`:

- `invalidate_config_context_for_objects()` (line 19) — the `pks = list(pks)` at line 29 evaluates a caller-supplied queryset on a routed connection, and the `chunked_update(Model.objects.filter(pk__in=pks), ...)` write at lines 34–38 is likewise unpinned. `transaction.on_commit()` at line 55 also defaults to the `default` alias.
- `invalidate_for_scope_delta()` (line 71) — the `Model.objects`, `TaggedItem.objects`, `Device.objects` and `VirtualMachine.objects` lookups it builds are all unpinned.
- `invalidate_config_context_for_configcontext()` (line 60) — passes querysets from `get_affected_objects()` straight through.

`extras/signals.py` — every receiver drops `using` into `**kwargs` and never reads it:

- `_make_object_save_handler._handler` (line 207) — no read, but the `chunked_update()` write downstream is still routed.
- `invalidate_on_device_vm_tag_change` (line 228) — `m2m_changed` supplies `using`.
- `_make_direct_upstream_handler._handler` (line 254) — the two failing call sites, lines 261 and 266.
- `_make_reparent_handler._handler` (line 275) — `model.objects` reads at lines 287 and 291, plus the `Device`/`VirtualMachine` lookups at lines 297 and 302.
- `_make_upstream_delete_handler._handler` (line 365) — `pre_delete` supplies `using`.

This is the same class of defect as #22922 and #22963, which were fixed on 4.6.x for the scope propagation and counter cache signals; the config-context cache subsystem landed on `feature` afterwards and was never given the same treatment.

### Proposed Fix

Thread the saving alias through the whole invalidation path, mirroring the #22922/#22963 fixes:

1. Give `invalidate_config_context_for_objects()`, `invalidate_for_scope_delta()` and `invalidate_config_context_for_configcontext()` a `using=None` keyword; pin every queryset they build with `.using(using)` (or `.db_manager(using)` for manager methods), and pass `using=using` to `transaction.on_commit()`.
2. Have each receiver in `extras/signals.py` read `using = kwargs.get('using')`, pin the `Device`/`VirtualMachine`/ltree lookups it builds, and forward the alias.
3. Add coverage alongside the existing `ScopeSignalConnectionTestCase` (`dcim/tests/test_signals.py`) and `VirtualizationSignalConnectionTestCase` (`virtualization/tests/test_signals.py`) using `PinnedConnectionRouter`.

Note that once fixed, `test_location_save_pins_queries_to_saving_connection` and `test_cluster_save_pins_vm_update_to_saving_connection` — which arrived on `feature` with the merge of `main` and currently fail for this reason — should pass without modification.

### PR DESCRIPTION
**Pull request by jeremystretch**
Closes: #23027

### Problem

The config-context cache invalidation added in #21025 (new in 4.7) issued every one of its queries unpinned, so Django resolved each through `DATABASE_ROUTERS` rather than using the alias the triggering signal supplies as `using`. Django consults the routers only for queries which name no connection, so nothing in the invalidation path ever saw the saving connection.

On a deployment with routers configured (e.g. `netbox_branching`), a router which resolved these elsewhere would read the affected PKs from — and NULL `_config_context_data` in — a database other than the one holding the change that triggered the invalidation. The intended objects keep a stale pre-rendered config context, and unrelated objects in the other database are needlessly invalidated. The write also fell outside the transaction the save runs in, so it was not rolled back with a failed save.

### Fix

Thread `using` through the invalidation path, following the pattern the scope propagation handlers in `dcim.signals` and `virtualization.signals` adopted in #22922 and #22963.

**`extras/cache.py`** — all three entry points take `using=None`:

* `invalidate_config_context_for_objects()` pins the `chunked_update()` NULL-out (which propagates `_db` to its keyset PK lookups and enclosing `transaction.atomic`), and ties `transaction.on_commit()` to the same connection.
* `invalidate_config_context_for_configcontext()` passes the alias through to `get_affected_objects()`.
* `invalidate_for_scope_delta()` pins the ltree `path` lookups, the `TaggedItem` lookups, and the Device/VM PK selects.

**`extras/models/configs.py`** — `get_affected_objects()` and `_get_affected_object_filters()` accept `using`, pinning the eagerly evaluated M2M scope reads (`_nested_scope_q`, `_direct_pks`) and returning querysets already bound to the alias.

**`extras/signals.py`** — every invalidation receiver accepts `using` and forwards it: the ConfigContext save/delete/M2M handlers, the Device/VM save and tag handlers, and the generated direct-upstream, reparent, and upstream-delete handlers. The reparent handler's own `path` and subtree reads are pinned as well.

Because the NULL-out now runs on the saving connection, it also lands inside the save's transaction and is rolled back with a failed save.

### Tests

Added `ConfigContextInvalidationRoutingTest` to `extras/tests/test_configcontext_cache.py`, using the existing `PinnedConnectionRouter` helper. 11 tests cover the cache helper called directly, `Location.site`, `Cluster.scope`, `Region` reparent, ConfigContext save and delete, scope M2M add/remove across both direct and nested (ltree) dimensions, an upstream `SET_NULL` delete, and a Device tag change.

* New tests: all 11 error on the unpatched code, all 11 pass with the fix.
* Regression run over `extras.tests.test_configcontext_cache`, `extras.tests.test_signals`, `extras.tests.test_models`, `extras.tests.test_jobs`, `dcim.tests.test_signals`, `virtualization.tests.test_signals`: 270 tests, OK.

### Out of scope

`RenderConfigContextJob`'s background sweep still routes its own queries. It runs outside any signal, so there is no triggering alias for it to inherit; scoping the sweep across databases would be a separate change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by jeremystretch** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by jeremystretch** [comment]
1. Only a concern when `using=None`, which does not occur in practice.
2. Extended the test coverage to address this.
3. Keeping as-is as a defensive measure.
4. Pre-existing issue; out of scope for this PR.
5. Out of scope for this PR.

File: netbox/extras/tests/test_configcontext_cache.py [hunk 1]
```diff
@@ -707,3 +713,250 @@ def test_single_query_serves_mixed_warm_and_cold(self):
...
+        location = Location.objects.get(pk=self.location.pk)
+        location.site_id = self.site_b.pk
+        with override_settings(DATABASE_ROUTERS=[PinnedConnectionRouter(Device)]):
+            location.save()
+
+    def test_cluster_scope_change_pins_vm_query(self):
+        cluster = Cluster.objects.get(pk=self.cluster.pk)
+        cluster.scope_id = self.site_b.pk
+        with override_settings(DATABASE_ROUTERS=[PinnedConnectionRouter(VirtualMachine)]):
+            cluster.save()
```
Comment (pheus) [inline netbox/extras/tests/test_configcontext_cache.py:822]: Could we also cover the normal snapshotted save path here? Calling `cluster.snapshot()`, populating the VM cache, and asserting that it is cleared would make this test exercise the watched-field check.

Without a snapshot, `_changed_fields()` returns `True` conservatively. Since `Cluster` stores its effective site in `_site_id`, the current `site_id` trigger appears to miss this case; I think it should watch `_site_id` instead.

File: netbox/extras/tests/test_configcontext_cache.py [hunk 2]
```diff
@@ -707,3 +713,250 @@ def test_single_query_serves_mixed_warm_and_cold(self):
...
+    DeviceRole,
+    DeviceType,
+    Location,
+    Platform,
+    Region,
+    Site,
+    SiteGroup,
+    Tenant,
+    TenantGroup,
+)
```
Comment (pheus) [inline netbox/extras/tests/test_configcontext_cache.py:754]: Could we include `Tag` here as well? `_get_affected_object_filters()` always evaluates the ConfigContext’s `tags` relation, so leaving it out means these tests would not catch an unpinned tag-scope lookup.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks for the thorough work!

I found one Cluster scope edge case that I think should be addressed before merging, and left one small non-blocking note about covering `Tag` in the routing matrix.

**Review by pheus — APPROVED** [review]
Thanks for the update. I didn’t find any further concerns. Looks good to me.
